### PR TITLE
Fix ARM32 tests

### PR DESCRIPTION
**Pull request by gergocs**
This patch fix every non SIMD test on ARM32

Signed-off-by: Gergo Csizi gergocs@inf.u-szeged.hu

### CODE REVIEW

File: src/jit/MemoryInl.h [hunk 1]
```diff
@@ -709,11 +709,7 @@ static void emitStore(sljit_compiler* compiler, Instruction* instr)
             addr.loadArg.argw = 0;
         }
 
-#if (defined SLJIT_FPU_UNALIGNED && SLJIT_FPU_UNALIGNED)
         sljit_emit_fop1(compiler, opcode, addr.memArg.arg, addr.memArg.argw, addr.loadArg.arg, addr.loadArg.argw);
-#else /* SLJIT_FPU_UNALIGNED */
-        sljit_emit_fmem(compiler, opcode | SLJIT_MEM_UNALIGNED, addr.loadArg.arg, addr.memArg.arg, addr.memArg.argw);
-#endif /* SLJIT_FPU_UNALIGNED */
```
Comment (zherczeg) [inline src/jit/MemoryInl.h:716]: Is this an optimization?

File: src/jit/IntMath32Inl.h [hunk 2]
```diff
@@ -611,14 +626,14 @@ static void emitBinary(sljit_compiler* compiler, Instruction* instr)
             opcode = SLJIT_XOR;
             break;
         case ByteCode::I32ShlOpcode:
-            opcode = SLJIT_SHL;
-            break;
+            emitShift32(compiler, SLJIT_SHL, args);
```
Comment (zherczeg) [inline src/jit/IntMath32Inl.h:629]: There is a masked variant of these opcodes, e.g. SLJIT_MSHL:

https://github.com/zherczeg/sljit/blob/master/sljit_src/sljitLir.h#L1181

**Review by zherczeg — CHANGES_REQUESTED** [review]

File: src/jit/MemoryInl.h (hunk 1 above)
Comment (gergocs) [reply src/jit/MemoryInl.h:716]: This is a fix. The previous implementation was bad.

File: src/jit/IntMath32Inl.h [hunk 3]
```diff
@@ -193,6 +193,22 @@ static void emitSimpleBinary64(sljit_compiler* compiler, sljit_s32 op1, sljit_s3
     sljit_emit_op2(compiler, op2, args[2].arg2, args[2].arg2w, args[0].arg2, args[0].arg2w, args[1].arg2, args[1].arg2w);
 }
 
+static void emitShift32(sljit_compiler* compiler, sljit_s32 op, JITArg* args)
+{
+    if (SLJIT_IS_IMM(args[1].arg) && args[1].argw == 0) {
+        if (args[0].arg != args[2].arg) {
```
Comment (zherczeg) [inline src/jit/IntMath32Inl.h:199]: Please also check argw as well.

File: src/jit/IntMath32Inl.h [hunk 4]
```diff
@@ -193,6 +193,22 @@ static void emitSimpleBinary64(sljit_compiler* compiler, sljit_s32 op1, sljit_s3
...
+            sljit_emit_op1(compiler, SLJIT_MOV, args[2].arg, args[2].argw, args[0].arg, args[0].argw);
+            return;
+        }
```
Comment (zherczeg) [inline src/jit/IntMath32Inl.h:202]: You can always return here, so no `else` is needed.

**Review by zherczeg — CHANGES_REQUESTED** [review]

File: src/jit/IntMath32Inl.h [hunk 5]
```diff
@@ -193,6 +193,23 @@ static void emitSimpleBinary64(sljit_compiler* compiler, sljit_s32 op1, sljit_s3
     sljit_emit_op2(compiler, op2, args[2].arg2, args[2].arg2w, args[0].arg2, args[0].arg2w, args[1].arg2, args[1].arg2w);
 }
 
+static void emitShift32(sljit_compiler* compiler, sljit_s32 op, JITArg* args)
+{
+    if (SLJIT_IS_IMM(args[1].arg) && args[1].argw == 0) {
+        if (args[0].arg != args[2].arg && args[0].argw != args[2].argw) {
```
Comment (zherczeg) [inline src/jit/IntMath32Inl.h:199]: && -> ||

**Pull request merged**